go/libraries/doltcore/sqle/dolt_session.go

```diff
@@ -79,11 +89,141 @@ func (sess *DoltSession) CommitTransaction(ctx *sql.Context) error {
 		return sql.ErrDatabaseNotFound.New(currentDb)
 	}
 
+	dbData := sess.dbDatas[currentDb]
+
 	root := dbRoot.root
-	h, err := dbRoot.ddb.WriteRootValue(ctx, root)
+	h, err := dbData.ddb.WriteRootValue(ctx, root)
+	if err != nil {
+		return err
+	}
+
+	return dbData.rsw.SetWorkingHash(ctx, h)
+}
+
+func (sess *DoltSession) GetDoltDB(dbName string) (*doltdb.DoltDB, bool) {
+	d, ok := sess.dbDatas[dbName]
+
+	if !ok {
+		return nil, false
+	}
+
+	return d.ddb, true
+}
+
+// GetRoot returns the current *RootValue for a given database associated with the session
+func (sess *DoltSession) GetRoot(dbName string) (*doltdb.RootValue, bool) {
+	dbRoot, ok := sess.dbRoots[dbName]
+
+	if !ok {
+		return nil, false
+	}
+
+	return dbRoot.root, true
+}
+
+func (sess *DoltSession) GetParent(ctx context.Context, dbName string) (*doltdb.Commit, error) {
+	dbd, dbFound := sess.dbDatas[dbName]
+
+	if !dbFound {
+		return nil, sql.ErrDatabaseNotFound.New(dbName)
+	}
+
+	_, value := sess.Session.Get(dbName + HeadKeySuffix)
+	valStr, isStr := value.(string)
+
+	if !isStr || !hash.IsValid(valStr) {
+		return nil, doltdb.ErrInvalidHash
+	}
+
+	cs, err := doltdb.NewCommitSpec(valStr, "")
+
+	if err != nil {
+		return nil, err
+	}
+
+	cm, err := dbd.ddb.Resolve(ctx, cs)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return cm, nil
+}
+
+func (sess *DoltSession) Set(ctx context.Context, key string, typ sql.Type, value interface{}) error {
+	if isHead, dbName := IsHeadKey(key); isHead {
+		dbd, dbFound := sess.dbDatas[dbName]
+
+		if !dbFound {
+			return sql.ErrDatabaseNotFound.New(dbName)
+		}
+
+		valStr, isStr := value.(string)
+
+		if !isStr || !hash.IsValid(valStr) {
+			return doltdb.ErrInvalidHash
+		}
+
+		cs, err := doltdb.NewCommitSpec(valStr, "")
+
+		if err != nil {
+			return err
+		}
+
+		cm, err := dbd.ddb.Resolve(ctx, cs)
+
+		if err != nil {
+			return err
+		}
+
+		root, err := cm.GetRootValue()
+
+		if err != nil {
+			return err
+		}
+
+		h, err := root.HashOf()
+
+		if err != nil {
+			return err
+		}
+
+		err = sess.Session.Set(ctx, key, typ, value)
+
+		if err != nil {
+			return err
+		}
+
+		hashStr := h.String()
+		err = sess.Session.Set(ctx, dbName+WorkingKeySuffix, sql.Text, hashStr)
+
+		if err != nil {
+			return err
+		}
+
+		sess.dbRoots[dbName] = dbRoot{hashStr, root}
+		return nil
+	}
+
+	return sess.Session.Set(ctx, key, typ, value)
+}
+
+func (sess *DoltSession) AddDB(ctx context.Context, name string, rsr env.RepoStateReader, rsw env.RepoStateWriter, ddb *doltdb.DoltDB) error {
```

Comment:
Any reason not to take a dsqle.Database insetead of all these params? Is there any use case where we aren't adding a Database object?

Reply:
done